src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java

```diff
@@ -323,36 +323,34 @@ private Stream<SourceFile> processMainSources(
         javaParserBuilder.classpath(dependencies).typeCache(typeCache);
         kotlinParserBuilder.classpath(dependencies).typeCache(new JavaTypeCache());
 
-        Stream<? extends SourceFile> cus = mainJavaSources.isEmpty() ? Stream.empty()
-                : Stream.of(javaParserBuilder).map(JavaParser.Builder::build)
-                        .flatMap(parser -> parser.parse(mainJavaSources, baseDir, ctx));
+        Stream<SourceFile> parsedJava = Stream.empty();
+        if (!mainJavaSources.isEmpty()) {
+            parsedJava = javaParserBuilder.build().parse(mainJavaSources, baseDir, ctx);
+            logDebug(mavenProject, "Scanned " + mainJavaSources.size() + " java source files in main scope.");
+        }
 
-        Stream<? extends SourceFile> kcus = mainKotlinSources.isEmpty() ? Stream.empty()
-                : Stream.of(kotlinParserBuilder).map(KotlinParser.Builder::build)
-                        .flatMap(parser -> parser.parse(mainKotlinSources, baseDir, ctx));
+        Stream<SourceFile> parsedKotlin = Stream.empty();
+        if (!mainKotlinSources.isEmpty()) {
+            parsedKotlin = kotlinParserBuilder.build().parse(mainKotlinSources, baseDir, ctx);
+            logDebug(mavenProject, "Scanned " + mainKotlinSources.size() + " kotlin source files in main scope.");
+        }
 
         List<Marker> mainProjectProvenance = new ArrayList<>(projectProvenance);
         mainProjectProvenance.add(sourceSet("main", dependencies, typeCache));
 
-        Stream<SourceFile> parsedJava = cus.map(addProvenance(baseDir, mainProjectProvenance, generatedSourcePaths));
-        logDebug(mavenProject, "Scanned " + mainJavaSources.size() + " java source files in main scope.");
-
-        Stream<SourceFile> parsedKotlin = kcus.map(addProvenance(baseDir, mainProjectProvenance, generatedSourcePaths));
-        logDebug(mavenProject, "Scanned " + mainKotlinSources.size() + " kotlin source files in main scope.");
-
         //Filter out any generated source files from the returned list, as we do not want to apply the recipe to the
         //generated files.
         Path buildDirectory = baseDir.relativize(Paths.get(mavenProject.getBuild().getDirectory()));
-        Stream<SourceFile> sourceFiles = Stream.concat(parsedJava, parsedKotlin).filter(s -> !s.getSourcePath().startsWith(buildDirectory));
+        Stream<SourceFile> sourceFiles = Stream.concat(parsedJava, parsedKotlin)
+                .filter(s -> !s.getSourcePath().startsWith(buildDirectory))
```

Comment:
Seems like it would be even better if we could filter out these generated sources before parsing them.

Reply:
I had thought we parse them such that we have their types available for attribution, as indicated on line 302 above. What improvement would you make here then?

Reply:
I suppose we have to test this, but these sources should still not need to be parsed, because the type attribution in other sources is via class files.

Reply:
I think it boils down to the question if `getCompileClasspathElements()` returns the classes compiled fom generated sources as well, which could depend on a number of different factors.
https://github.com/openrewrite/rewrite-maven-plugin/blob/0b8fc330f355a96b6e5b7a103601c07f775d9de3/src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java#L317-L324

Right now we define generated sources as anything we find in the `target/` directory, and parse those as source rather than compiled classes.
https://github.com/openrewrite/rewrite-maven-plugin/blob/0b8fc330f355a96b6e5b7a103601c07f775d9de3/src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java#L302-L308

While I agree it might be interesting to not parse those generated sources as source files but rather only as classpath entries, I'm not entirely sure that would work in all cases, and verifying that would take more time than I currently have available. Are you ok merging this PR as-is to at least simplify the logic?